src/protocol/sort/multi_bit_permutation.rs

```diff
@@ -0,0 +1,256 @@
+        equality_check_futures
+            .push(async move { get_bit_equality_checkers(j, input, ctx_equality_checker).await });
+    }
+
+    let equality_checks = try_join_all(equality_check_futures).await?;
+
+    // Compute accumulated sum
+    let mut sop_inputs_transposed = Vec::with_capacity(num_possible_bit_values * num_records);
+    let mut initial_state = S::ZERO;
+    for equality_check in equality_checks {
+        let out = equality_check
+            .into_iter()
+            .scan(initial_state, |sum, x| {
+                *sum += &x;
+                Some((x, sum.clone()))
+            })
+            .collect::<Vec<_>>();
+        initial_state = out
+            .last()
+            .unwrap_or_else(|| panic!("We should never reach here"))
+            .1
+            .clone();
+        sop_inputs_transposed.push(out);
+    }
+
+    // Take sum of products of output of equality check and accumulated sum
+    let mut permutation_futures = Vec::with_capacity(num_records);
+    for rec in 0..num_records {
+        let ctx_sop = ctx.narrow(&Sop);
+        let mut sop_inputs = Vec::with_capacity(num_possible_bit_values);
+        for sop_input_transposed in sop_inputs_transposed.iter().take(num_possible_bit_values) {
+            sop_inputs.push((&sop_input_transposed[rec].0, &sop_input_transposed[rec].1));
+        }
+        permutation_futures.push(async move {
+            ctx_sop
+                .sum_of_products(RecordId::from(rec), sop_inputs.as_slice())
+                .await
+        });
+    }
+    try_join_all(permutation_futures).await
+}
+
+/// For a given `idx` check if each of the record has same value as idx.
+/// Steps
+/// 1. Get bit representation of `idx`
+/// 2. Calculate equality check
+///   i. keep record bit if `idx` bit is 1
+///   ii.toggle record bit if idx bit is 0 (Done by taking `share_of_one` - value)
+/// 3. Multiply equality checks for all bits per record - in clear per record this will be 1 only for 1 index while 0 for all others
+async fn get_bit_equality_checkers<F, C, S>(
+    idx: usize,
+    input: &[Vec<S>],
+    ctx: C,
+) -> Result<Vec<S>, Error>
+where
+    F: Field,
+    C: Context<F, Share = S>,
+    S: SecretSharing<F>,
+{
+    let num_multi_bits = input.len();
+    let num_records = input[0].len();
+
+    let mut equality_check_futures = Vec::with_capacity(num_records);
+
+    let j_bit_representation = get_binary_from_int(idx, num_multi_bits);
+
+    for rec in 0..num_records {
+        let ctx_across_bits = ctx.narrow(&MultiplyAcrossBits);
+        let share_of_one = ctx_across_bits.share_of_one();
+
+        let mut bits_for_record = Vec::with_capacity(num_multi_bits);
+        for jth_bits in input.iter().take(num_multi_bits) {
+            bits_for_record.push(jth_bits[rec].clone());
+        }
+        let mut mult_input = Vec::with_capacity(num_multi_bits);
+        for jth_bit in j_bit_representation.iter().take(num_multi_bits) {
```

Comment:
I feel like you should be able to zip `input` and `j_bits` like this

```rust
   input.iter().zip(j_bits).map(|(record, jth_bit)| { if jth_bit { record[rec] } else { -record[rec] + share_of_one } })
```

Reply:
Sounds good.. made this change